lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -2781,12 +2851,17 @@ ConformanceChecker::resolveWitnessViaLookup(ValueDecl *requirement) {
           auto ctor = cast<ConstructorDecl>(requirement);
           auto witnessCtor = cast<ConstructorDecl>(witness);
           auto &diags = witness->getASTContext().Diags;
-          diags.diagnose(witness,
+          SourceLoc diagLoc = getLocForDiagnosingWitness(conformance, witness);
+          diags.diagnose(diagLoc,
                          diag::witness_initializer_failability,
                          ctor->getFullName(),
                          witnessCtor->getFailability()
                            == OTK_ImplicitlyUnwrappedOptional)
             .highlight(witnessCtor->getFailabilityLoc());
+          if (diagLoc != witness->getLoc()) {
```

Comment:
It seems every second diagnostic touched here gets this `if` block almost verbatim. We could consider factoring it out somehow?

Reply:
I guess that makes sense for all the `decl_declared_here` ones, at least. I tried to think of how to fold it into `getLocForDiagnosingWitness` with a callback and couldn't, then tried to think of how to generalize it over any possible follow-up note and couldn't. But more of them ended up just being "declared here" than I originally thought.